app/views.py

```diff
@@ -0,0 +1,451 @@
+from flask_login import login_required, login_user, logout_user, current_user
+from app.search import YTSearch, LastFMSearch
+from flask import current_app as app
+from sqlalchemy import exc, or_, select
+from .pagination import pagination
+import app.forms
+
+main = Blueprint("main", __name__)
+
+login_manager.login_view = "main.login"
+login_manager.login_message_category = "danger"
+# login_manager.session_protection = app.config["SESSION_PROTECTION"]
+
+
+@login_manager.user_loader
+def load_user(user_id):
+    try:
+        return User.query.get(user_id)
+    except exc.SQLAlchemyError:
+        return None
+
+
+@main.route("/login", methods=["GET", "POST"])
+def login():
+    form = app.forms.LoginForm()
+    if form.validate_on_submit():
+        username = form.username.data
+        if User.query.filter_by(username=username).first():
+            user = User.authenticate(form.username.data, form.password.data)
+            if user:
+                if form.remember.data:
+                    login_user(user, remember=True)
+                else:
+                    login_user(user)
+                flash(f"Welcome back {username}", "success")
+
+                next = request.args.get("next")
+                # if not is_safe_url(next):
+                #     return abort(400)
+                # need to implement is_safe_url function for redirects
+                return redirect(next or url_for(".root"))
+            flash("Invalid credentials", "danger")
+        else:
+            flash("Username not found", "danger")
+    return render_template("login.html", form=form)
+
+
+@main.route("/logout")
+def logout():
+    logout_user()
+    return redirect(url_for(".root"))
+
+
+@main.route("/signup", methods=["GET", "POST"])
+def signup():
+    form = app.forms.SignupForm()
+
+    if form.validate_on_submit():
+        try:
+            if form.avatar.data == "":
+                avatar = None
+            else:
+                avatar = form.avatar.data
+            new_user = User.signup(
+                username=form.username.data,
+                password=form.password.data,
+                email=form.email.data,
+                avatar=avatar,
+                bio=form.bio.data,
+            )
+            db.session.commit()
+            flash("Welcome to Check Out This Song!", "success")
+
+        except exc.IntegrityError as error:
+            if "username" in str(error.orig):
+                flash("That username is already taken!", "danger")
+            if "email" in str(error.orig):
+                flash(
+                    "An account already exists for that email address please login",
+                    "warning",
+                )
+
+            return render_template("signup.html", form=form)
+
+        login_user(new_user)
+        return redirect(url_for(".root"))
+
+    else:
+        return render_template("signup.html", form=form)
+
+
+@main.route("/")
+def root():
+    db.session.rollback()
```

Comment:
Why are you doing this here? Don't think this is a good idea. Flask-SQLAlchemy or similar should be able to better handle the session lifecycle per request